src/trio/_threads.py

```diff
@@ -52,6 +54,72 @@ class _ParentTaskData(threading.local):
 _thread_counter = count()
 
 
+@define
+class _ActiveThreadCount:
+    count: int
+    event: Event
+
+
+_active_threads_local: RunVar[_ActiveThreadCount] = RunVar("active_threads")
+
+
+@contextlib.contextmanager
+def _track_active_thread() -> Generator[None, None, None]:
+    try:
+        active_threads_local = _active_threads_local.get()
+    except LookupError:
+        active_threads_local = _ActiveThreadCount(0, Event())
+        _active_threads_local.set(active_threads_local)
+
+    active_threads_local.count += 1
+    try:
+        yield
+    finally:
+        active_threads_local.count -= 1
+        if active_threads_local.count == 0:
+            active_threads_local.event.set()
+            active_threads_local.event = Event()
+
+
+async def wait_all_threads_completed() -> None:
+    """Wait until no threads are still running tasks.
+
+    This is intended to be used when testing code with trio.to_thread to
+    make sure no tasks are still making progress in a thread. See the
+    following code for a usage example::
+
+        async def wait_all_settled():
+            while True:
+                await trio.testing.wait_all_threads_complete()
+                await trio.testing.wait_all_tasks_blocked()
+                if trio.testing.active_thread_count() == 0:
+                    break
+    """
+
+    await checkpoint()
+
+    try:
+        active_threads_local = _active_threads_local.get()
+    except LookupError:
+        # If there would have been active threads, the
+        # _active_threads_local would have been set
+        return
+
+    while active_threads_local.count != 0:
+        await active_threads_local.event.wait()
```

Comment:
Just wanting to be sure, I know event wait call is doing what we want and will block until thread count is zero and it's set, but is there any way with race conditions that `active_threads_local.event` is reset to a new lock before the `wait` call sees that? Would it be beneficial to move the event resetting to before the `active_threads_local.count += 1` line?

Reply:
ooh. Yeah shouldn't it maybe be something like this:
```python
    try:
        active_threads_local = _active_threads_local.get()
    except LookupError:
        active_threads_local = _ActiveThreadCount(1, Event())
        _active_threads_local.set(active_threads_local)
    else:
        active_threads_local.count += 1
```

Reply:
Yea, probably something like that and adding a part in the else block there where it resets the event object if the event it has has already fired.

Reply:
I don't understand where a race could arise here. This is all happening in the main thread so everything between awaits is effectively atomic.

I have similar code in a different project and it hasn't given me trouble... Although speaking of that, it would be good to `assert active_threads_local.count >= 0` after decrementing it, because if a negative number sneaks in it'd be better to fail fast.

Reply:
Yeah I don't think we have to care about thread safety here. Rather, the only times `_active_threads_local` can be changed (that we worry about) is at an `await` point.

After all `to_thread_run_sync` can be only run from the main thread (that has your `trio` event loop).